barbican/keysource.go

```diff
@@ -0,0 +1,282 @@
+}
+
+// NewMasterKeyFromSecretHref takes a Barbican Secret Href and returns a new MasterKey for that
+func NewMasterKeyFromSecretHref(secretHref string) *MasterKey {
+	k := &MasterKey{}
+	k.SecretHref = secretHref
+	k.CreationDate = time.Now().UTC()
+	return k
+}
+
+// MasterKeysFromSecretHref takes a comma separated list of Secret Hrefs and returns a slice of new MasterKeys for them
+func MasterKeysFromSecretHref(secretHref string) []*MasterKey {
+	var keys []*MasterKey
+	if secretHref == "" {
+		return keys
+	}
+	for _, s := range strings.Split(secretHref, ",") {
+		keys = append(keys, NewMasterKeyFromSecretHref(s))
+	}
+	return keys
+}
+
+// ToMap converts the MasterKey to a map for serialization purposes
+func (key MasterKey) ToMap() map[string]interface{} {
+	out := make(map[string]interface{})
+	out["secret_href"] = key.SecretHref
+	out["enc"] = key.EncryptedKey
+	out["created_at"] = key.CreationDate.UTC().Format(time.RFC3339)
+	return out
+}
+
+func newKeyManager() (*gophercloud.ServiceClient, error) {
+
+	home, err := user.Current()
+	if err != nil {
+		log.WithError(err).Error("Failed to determine user home")
+		return nil, err
+	}
+
+	// support loading config gophercloud style
+	cfg, err := ini.LooseLoad(os.Getenv("GOPHERCLOUD_CONFIG"), fmt.Sprintf("%v/.cloud-config", home), "/etc/cloud-config")
+	if err == nil {
+		for _, k := range cfg.Section("Global").KeyStrings() {
+			// we're iterating keys, safe not to check for err
+			kv, _ := cfg.Section("Global").GetKey(k)
+			envVar := fmt.Sprintf("OS_%s", strings.ToUpper(strings.ReplaceAll(k, "-", "_")))
+			os.Setenv(envVar, kv.Value())
+		}
+	} else if !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	// fetch the client based on the env
+	var provider *gophercloud.ProviderClient
+
+	opts, err := clientconfig.AuthOptions(nil)
+	if err != nil {
+		log.WithError(err).Error("Failed to fetch auth options")
+		return nil, err
+	}
+	trustID, present := os.LookupEnv("OS_TRUST_ID")
+	if present {
+		optsExt := trusts.AuthOptsExt{
+			TrustID: trustID,
+			AuthOptionsBuilder: &tokens.AuthOptions{
+				IdentityEndpoint: opts.IdentityEndpoint,
+				UserID:           os.Getenv("OS_USER_ID"),
+				Password:         os.Getenv("OS_PASSWORD"),
+			},
+		}
+		//FIXME(rochaporto): this is silly, but it's easy to get a client like this
+		//we authenticate twice though
+		provider, err = openstack.AuthenticatedClient(*opts)
+		if err != nil {
+			log.WithError(err).Error("failed to get openstack client")
+			return nil, err
+		}
+		err = openstack.AuthenticateV3(provider, optsExt, gophercloud.EndpointOpts{})
+		if err != nil {
+			log.WithError(err).Error("failed to authenticate with v3")
+			return nil, err
+		}
+	} else {
+		provider, err = openstack.AuthenticatedClient(*opts)
+		if err != nil {
+			log.WithError(err).Error("failed to create authenticated client")
+			return nil, err
+		}
+	}
+
+	client, err := openstack.NewKeyManagerV1(provider,
+		gophercloud.EndpointOpts{Region: "cern"})
```

Comment:
@rochaporto This breaks usage outside of cern ;)
As the region is hardcoded and not using env variables, it results in

```bash
[BARBICAN]       ERRO[0000] failed to create key manager                  error="No suitable endpoint could be found in the service catalog."
[BARBICAN]       ERRO[0000] Encryption failed                             SecretHref="[REMOVED]"
Could not generate data key: [failed to encrypt new data key with master key "[REMOVED]": Cannot create KeyManager service: No suitable endpoint could be found in the service catalog.]
```

Error message

Reply:
I agree. It would be better to source it from the OS_REGION_NAME and pass it to gophercloud like it was done with all other config vars.

https://github.com/mozilla/sops/pull/683/commits/e8d541db0ce426778ad8740e29c4984f82bbafcc#diff-772389010f5fe2d4bd05f8aed0cc0a4dd204857414fdc83611067b7aa73045adR162-R168 #